File: platforms/macos/PixCapMac/Sources/PixCapMac/Editor/AnnotationModel.swift
import Foundation
import AppKit
import SwiftUI

/// Tools available in the annotation toolbar.
///
/// Coordinates for every annotation are stored in *image space*: origin at the
/// top-left of the captured screenshot, measured in image points. That keeps the
/// model independent of preview zoom, so the same items render identically in
/// the on-screen preview and in the full-resolution export.
public enum AnnotationTool: String, CaseIterable, Identifiable {
    case select
    case crop
    case arrow
    case line
    case rectangle
    case ellipse
    case freehand
    case highlight
    case text
    case counter
    case blur
    case redaction
    case spotlight

    public var id: String { rawValue }

    public var symbolName: String {
        switch self {
        case .select: return "cursorarrow"
        case .crop: return "crop"
        case .arrow: return "arrow.up.right"
        case .line: return "line.diagonal"
        case .rectangle: return "rectangle"
        case .ellipse: return "circle"
        case .freehand: return "scribble"
        case .highlight: return "highlighter"
        case .text: return "textformat"
        case .counter: return "number.circle"
        case .blur: return "drop.halffull"
        case .redaction: return "rectangle.fill"
        case .spotlight: return "light.max"
        }
    }

    public var title: String {
        switch self {
        case .select: return "Select"
        case .crop: return "Crop"
        case .arrow: return "Arrow"
        case .line: return "Line"
        case .rectangle: return "Rectangle"
        case .ellipse: return "Ellipse"
        case .freehand: return "Draw"
        case .highlight: return "Highlight"
        case .text: return "Text"
        case .counter: return "Counter"
        case .blur: return "Blur"
        case .redaction: return "Redaction"
        case .spotlight: return "Spotlight"
        }
    }

    /// Single-key shortcut, matching the shortcuts listed in Preferences.
    public var shortcut: String {
        switch self {
        case .select: return "V"
        case .crop: return "K"
        case .arrow: return "A"
        case .line: return "L"
        case .rectangle: return "R"
        case .ellipse: return "E"
        case .freehand: return "D"
        case .highlight: return "M"
        case .text: return "T"
        case .counter: return "C"
        case .blur: return "B"
        case .redaction: return "P"
        case .spotlight: return "H"
        }
    }

    /// Tools that are placed with a single click rather than dragged out.
    public var isPointPlaced: Bool {
        self == .counter || self == .text
    }

    /// Tools that modify the canvas rather than adding an item to the layer.
    public var isCanvasOperation: Bool {
        self == .crop
    }
}

public enum BlurStyle: String, CaseIterable, Codable {
    case gaussian = "Blur"
    case pixelate = "Pixelate"
}

public enum ArrowHead: String, CaseIterable, Codable {
    case filled
    case open
    case none

    public var title: String {
        switch self {
        case .filled: return "Filled"
        case .open: return "Open"
        case .none: return "None"
        }
    }
}

/// One drawn annotation.
public struct AnnotationItem: Identifiable, Equatable {
    public let id: UUID
    public var tool: AnnotationTool
    public var start: CGPoint
    public var end: CGPoint
    public var points: [CGPoint]
    public var colorHex: String
    public var strokeWidth: CGFloat
    public var filled: Bool
    public var text: String
    public var fontSize: CGFloat
    public var number: Int
    public var blurStyle: BlurStyle
    public var blurIntensity: CGFloat
    public var curvedArrow: Bool
    public var arrowHead: ArrowHead
    public var dashed: Bool
    /// How far the area outside a spotlight is dimmed (0–1).
    public var spotlightDim: CGFloat

    public init(
        id: UUID = UUID(),
        tool: AnnotationTool,
        start: CGPoint,
        end: CGPoint,
        points: [CGPoint] = [],
        colorHex: String,
        strokeWidth: CGFloat,
        filled: Bool = false,
        text: String = "",
        fontSize: CGFloat = 24,
        number: Int = 1,
        blurStyle: BlurStyle = .gaussian,
        blurIntensity: CGFloat = 20,
        curvedArrow: Bool = true,
        arrowHead: ArrowHead = .filled,
        dashed: Bool = false,
        spotlightDim: CGFloat = 0.65
    ) {
        self.id = id
        self.tool = tool
        self.start = start
        self.end = end
        self.points = points
        self.colorHex = colorHex
        self.strokeWidth = strokeWidth
        self.filled = filled
        self.text = text
        self.fontSize = fontSize
        self.number = number
        self.blurStyle = blurStyle
        self.blurIntensity = blurIntensity
        self.curvedArrow = curvedArrow
        self.arrowHead = arrowHead
        self.dashed = dashed
        self.spotlightDim = spotlightDim
    }

    public var color: NSColor {
        NSColor(hex: colorHex) ?? .systemRed
    }

    /// Normalised bounding rectangle in image space.
    public var rect: CGRect {
        CGRect(
            x: min(start.x, end.x),
            y: min(start.y, end.y),
            width: abs(end.x - start.x),
            height: abs(end.y - start.y)
        )
    }

    /// Whether the item covers enough area to be worth keeping after a drag.
    public var isMeaningful: Bool {
        switch tool {
        case .counter:
            return true
        case .text:
            return !text.isEmpty
        case .freehand:
            return points.count > 1
        case .line, .arrow:
            return hypot(end.x - start.x, end.y - start.y) > 4
        default:
            return rect.width > 4 && rect.height > 4
        }
    }

    /// Hit-test area used by the select tool, padded so thin strokes stay grabbable.
    public func hitRect() -> CGRect {
        let padding = max(strokeWidth, 8)
        switch tool {
        case .counter:
            let radius = counterRadius
            return CGRect(x: start.x - radius, y: start.y - radius, width: radius * 2, height: radius * 2)
        case .text:
            let size = textSize()
            return CGRect(x: start.x, y: start.y, width: size.width, height: size.height).insetBy(dx: -padding, dy: -padding)
        case .freehand:
            guard let first = points.first else { return .zero }
            var bounds = CGRect(origin: first, size: .zero)
            for point in points.dropFirst() {
                bounds = bounds.union(CGRect(origin: point, size: .zero))
            }
            return bounds.insetBy(dx: -padding, dy: -padding)
        default:
            return rect.insetBy(dx: -padding, dy: -padding)
        }
    }

    public var counterRadius: CGFloat { max(14, strokeWidth * 6) }

    public func textAttributes() -> [NSAttributedString.Key: Any] {
        [
            .font: NSFont.systemFont(ofSize: fontSize, weight: .semibold),
            .foregroundColor: color
        ]
    }

    public func textSize() -> CGSize {
        let string = text.isEmpty ? " " : text
        return (string as NSString).size(withAttributes: textAttributes())
    }

    /// Returns a copy translated by `delta` (used by the select tool).
    public func moved(by delta: CGSize) -> AnnotationItem {
        var copy = self
        copy.start = CGPoint(x: start.x + delta.width, y: start.y + delta.height)
        copy.end = CGPoint(x: end.x + delta.width, y: end.y + delta.height)
        copy.points = points.map { CGPoint(x: $0.x + delta.width, y: $0.y + delta.height) }
        return copy
    }
}

/// Holds the annotation list plus undo/redo history for one editor session.
public final class AnnotationStore: ObservableObject {
    @Published public private(set) var items: [AnnotationItem] = []
    @Published public var selectedID: UUID?
    /// Item currently being typed into by the text tool.
    @Published public var editingID: UUID?

    private var undoStack: [[AnnotationItem]] = []
    private var redoStack: [[AnnotationItem]] = []

    public init() {}

    public var canUndo: Bool { !undoStack.isEmpty }
    public var canRedo: Bool { !redoStack.isEmpty }

    /// The next counter value, continuing from the highest bubble already placed.
    public var nextCounterNumber: Int {
        let start = max(1, Settings.int(SettingsKey.counterStartNumber))
        let highest = items.filter { $0.tool == .counter }.map(\.number).max()
        return highest.map { $0 + 1 } ?? start
    }

    public func add(_ item: AnnotationItem) {
        checkpoint()
        items.append(item)
    }

    public func update(id: UUID, checkpointFirst: Bool = true, _ mutate: (inout AnnotationItem) -> Void) {
        guard let index = items.firstIndex(where: { $0.id == id }) else { return }
        if checkpointFirst { checkpoint() }
        mutate(&items[index])
    }

    public func remove(id: UUID) {
        guard items.contains(where: { $0.id == id }) else { return }
        checkpoint()
        items.removeAll { $0.id == id }
        if selectedID == id { selectedID = nil }
        if editingID == id { editingID = nil }
    }

    public func item(withID id: UUID) -> AnnotationItem? {
        items.first { $0.id == id }
    }

    /// Topmost item whose hit area contains `point` (image space).
    public func hitTest(_ point: CGPoint) -> AnnotationItem? {
        items.reversed().first { $0.hitRect().contains(point) }
    }

    public func clear() {
        guard !items.isEmpty else { return }
        checkpoint()
        items.removeAll()
        selectedID = nil
        editingID = nil
    }

    /// Replaces the whole layer, e.g. when loading a saved document.
    public func replaceAll(with newItems: [AnnotationItem], resetHistory: Bool = true) {
        if resetHistory {
            undoStack.removeAll()
            redoStack.removeAll()
        } else {
            checkpoint()
        }
        items = newItems
        selectedID = nil
        editingID = nil
    }

    public func undo() {
        guard let previous = undoStack.popLast() else { return }
        redoStack.append(items)
        items = previous
        pruneSelection()
    }

    public func redo() {
        guard let next = redoStack.popLast() else { return }
        undoStack.append(items)
        items = next
        pruneSelection()
    }

    /// Snapshots current state for undo. Call before any mutation.
    public func checkpoint() {
        undoStack.append(items)
        if undoStack.count > 50 { undoStack.removeFirst() }
        redoStack.removeAll()
    }

    private func pruneSelection() {
        if let selectedID, !items.contains(where: { $0.id == selectedID }) {
            self.selectedID = nil
        }
        if let editingID, !items.contains(where: { $0.id == editingID }) {
            self.editingID = nil
        }
    }
}

File: platforms/windows/PixCapWin/Views/AnnotationModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json.Nodes;
using Windows.Foundation;

namespace PixCapWin.Views;

/// <summary>
/// Tools the annotation editor offers.
///
/// Names match the strings the Rust renderer expects, so an item serialises
/// straight into the shared document schema without translation.
/// </summary>
public enum AnnotationTool
{
    Select,
    Crop,
    Rectangle,
    Ellipse,
    Line,
    Arrow,
    Freehand,
    Highlight,
    Text,
    Counter,
    Blur,
    Redaction,
    Spotlight
}

public static class AnnotationToolExtensions
{
    /// <summary>The identifier the renderer uses.</summary>
    public static string ToToken(this AnnotationTool tool) => tool switch
    {
        AnnotationTool.Select => "select",
        AnnotationTool.Crop => "crop",
        AnnotationTool.Rectangle => "rectangle",
        AnnotationTool.Ellipse => "ellipse",
        AnnotationTool.Line => "line",
        AnnotationTool.Arrow => "arrow",
        AnnotationTool.Freehand => "freehand",
        AnnotationTool.Highlight => "highlight",
        AnnotationTool.Text => "text",
        AnnotationTool.Counter => "counter",
        AnnotationTool.Blur => "blur",
        AnnotationTool.Redaction => "redaction",
        AnnotationTool.Spotlight => "spotlight",
        _ => "rectangle"
    };

    /// <summary>
    /// Short palette label. Now the tooltip and the accessible name for the
    /// tool's button, since the button itself shows <see cref="Glyph"/>.
    /// </summary>
    public static string Label(this AnnotationTool tool) => tool switch
    {
        AnnotationTool.Select => "Select",
        AnnotationTool.Crop => "Crop",
        AnnotationTool.Rectangle => "Rectangle",
        AnnotationTool.Ellipse => "Ellipse",
        AnnotationTool.Line => "Line",
        AnnotationTool.Arrow => "Arrow",
        AnnotationTool.Freehand => "Draw",
        AnnotationTool.Highlight => "Highlight",
        AnnotationTool.Text => "Text",
        AnnotationTool.Counter => "Counter",
        AnnotationTool.Blur => "Blur",
        AnnotationTool.Redaction => "Redaction",
        AnnotationTool.Spotlight => "Spotlight",
        _ => tool.ToString()
    };

    /// <summary>
    /// Segoe Fluent Icons code point for the palette button.
    ///
    /// Every one of these was rendered and looked at before being chosen — a
    /// glyph that turns out to be the wrong picture is worse than a word, and
    /// the code point alone does not tell you which picture you get. The label
    /// stays as the tooltip and the accessible name, so nothing is lost to
    /// someone who does not recognise an icon.
    /// </summary>
    public static string Glyph(this AnnotationTool tool) => tool switch
    {
        AnnotationTool.Select => "\uE8B0",      // pointer
        AnnotationTool.Crop => "\uE7A8",        // crop marks
        AnnotationTool.Rectangle => "\uE739",   // empty square
        AnnotationTool.Ellipse => "\uEA3A",     // empty circle
        AnnotationTool.Line => "\uE738",        // horizontal rule
        AnnotationTool.Arrow => "\uE72A",       // arrow
        AnnotationTool.Freehand => "\uEC87",    // pencil drawing a stroke
        AnnotationTool.Highlight => "\uE7E6",   // highlighter
        AnnotationTool.Text => "\uE97F",        // letter A
        AnnotationTool.Counter => "\uF146",     // numeral 1
        AnnotationTool.Blur => "\uEB42",        // droplet
        AnnotationTool.Redaction => "\uE73B",   // filled block
        AnnotationTool.Spotlight => "\uEC8A",   // sun
        _ => "\uE739"
    };

    /// <summary>Tools placed with a single click rather than dragged out.</summary>
    public static bool IsPointPlaced(this AnnotationTool tool) =>
        tool is AnnotationTool.Text or AnnotationTool.Counter;
}

/// <summary>
/// One annotation, in image-space points with a top-left origin — the same
/// coordinate system the renderer and the macOS editor use, so a document
/// written on either platform means the same thing.
/// </summary>
public sealed class AnnotationItem
{
    public Guid Id { get; init; } = Guid.NewGuid();
    public AnnotationTool Tool { get; set; }
    public Point Start { get; set; }
    public Point End { get; set; }
    public List<Point> Points { get; set; } = [];
    public string ColorHex { get; set; } = "#FF3366";
    public double StrokeWidth { get; set; } = 4;
    public bool Filled { get; set; }
    public string Text { get; set; } = "";
    public double FontSize { get; set; } = 28;
    public int Number { get; set; } = 1;
    public string BlurStyle { get; set; } = "pixelate";
    public double BlurIntensity { get; set; } = 24;
    public bool CurvedArrow { get; set; } = true;
    public string ArrowHead { get; set; } = "filled";
    public bool Dashed { get; set; }
    public double SpotlightDim { get; set; } = 0.65;

    /// <summary>Normalised bounding rectangle.</summary>
    public Rect Bounds => new(
        Math.Min(Start.X, End.X),
        Math.Min(Start.Y, End.Y),
        Math.Abs(End.X - Start.X),
        Math.Abs(End.Y - Start.Y));

    /// <summary>Whether a drag produced something worth keeping.</summary>
    public bool IsMeaningful => Tool switch
    {
        AnnotationTool.Counter => true,
        AnnotationTool.Text => !string.IsNullOrWhiteSpace(Text),
        AnnotationTool.Freehand => Points.Count > 1,
        AnnotationTool.Line or AnnotationTool.Arrow =>
            Math.Sqrt(Math.Pow(End.X - Start.X, 2) + Math.Pow(End.Y - Start.Y, 2)) > 4,
        _ => Bounds.Width > 4 && Bounds.Height > 4
    };

    /// <summary>Hit area for the select tool, padded so thin strokes stay grabbable.</summary>
    public Rect HitBounds()
    {
        var padding = Math.Max(StrokeWidth, 8);

        if (Tool == AnnotationTool.Counter)
        {
            var radius = Math.Max(14, StrokeWidth * 6);
            return new Rect(Start.X - radius, Start.Y - radius, radius * 2, radius * 2);
        }

        if (Tool == AnnotationTool.Text)
        {
            // Approximate: the renderer owns the real metrics.
            var width = Math.Max(20, Text.Length * FontSize * 0.55);
            return new Rect(Start.X - padding, Start.Y - padding,
                            width + padding * 2, FontSize + padding * 2);
        }

        if (Tool == AnnotationTool.Freehand && Points.Count > 0)
        {
            var minX = Points.Min(p => p.X);
            var minY = Points.Min(p => p.Y);
            var maxX = Points.Max(p => p.X);
            var maxY = Points.Max(p => p.Y);
            return new Rect(minX - padding, minY - padding,
                            maxX - minX + padding * 2, maxY - minY + padding * 2);
        }

        var bounds = Bounds;
        return new Rect(bounds.X - padding, bounds.Y - padding,
                        bounds.Width + padding * 2, bounds.Height + padding * 2);
    }

    public AnnotationItem Clone() => new()
    {
        Id = Id,
        Tool = Tool, Start = Start, End = End, Points = [.. Points],
        ColorHex = ColorHex, StrokeWidth = StrokeWidth, Filled = Filled,
        Text = Text, FontSize = FontSize, Number = Number,
        BlurStyle = BlurStyle, BlurIntensity = BlurIntensity,
        CurvedArrow = CurvedArrow, ArrowHead = ArrowHead, Dashed = Dashed,
        SpotlightDim = SpotlightDim
    };

    public AnnotationItem MovedBy(double dx, double dy)
    {
        var copy = Clone();
        copy.Start = new Point(Start.X + dx, Start.Y + dy);
        copy.End = new Point(End.X + dx, End.Y + dy);
        copy.Points = Points.Select(p => new Point(p.X + dx, p.Y + dy)).ToList();
        return copy;
    }

    /// <summary>Serialises into the shared document schema.</summary>
    public JsonObject ToJson() => new()
    {
        ["id"] = Id.ToString(),
        ["tool"] = Tool.ToToken(),
        ["start"] = new JsonArray(Start.X, Start.Y),
        ["end"] = new JsonArray(End.X, End.Y),
        ["points"] = new JsonArray(Points.Select(p => (JsonNode)new JsonArray(p.X, p.Y)).ToArray()),
        ["color_hex"] = ColorHex,
        ["stroke_width"] = StrokeWidth,
        ["filled"] = Filled,
        ["text"] = Text,
        ["font_size"] = FontSize,
        ["number"] = Number,
        ["blur_style"] = BlurStyle,
        ["blur_intensity"] = BlurIntensity,
        ["curved_arrow"] = CurvedArrow,
        ["arrow_head"] = ArrowHead,
        ["dashed"] = Dashed,
        ["spotlight_dim"] = SpotlightDim
    };
}

/// <summary>
/// The annotation layer plus undo/redo history for one editing session.
/// </summary>
public sealed class AnnotationStore
{
    private readonly List<AnnotationItem> _items = [];
    private readonly Stack<List<AnnotationItem>> _undo = new();
    private readonly Stack<List<AnnotationItem>> _redo = new();

    public IReadOnlyList<AnnotationItem> Items => _items;
    public Guid? SelectedId { get; set; }

    public bool CanUndo => _undo.Count > 0;
    public bool CanRedo => _redo.Count > 0;
    public bool IsEmpty => _items.Count == 0;

    /// <summary>Next counter value, continuing from the highest already placed.</summary>
    public int NextCounter
    {
        get
        {
            var placed = _items.Where(item => item.Tool == AnnotationTool.Counter).ToList();
            return placed.Count == 0 ? 1 : placed.Max(item => item.Number) + 1;
        }
    }

    /// <summary>Snapshots current state for undo. Call before any mutation.</summary>
    public void Checkpoint()
    {
        _undo.Push(_items.Select(item => item.Clone()).ToList());
        if (_undo.Count > 50)
        {
            // Stack has no bounded variant; rebuild without the oldest entry.
            var trimmed = _undo.Reverse().Skip(1).ToList();
            _undo.Clear();
            foreach (var state in trimmed) _undo.Push(state);
        }
        _redo.Clear();
    }

    public void Add(AnnotationItem item)
    {
        Checkpoint();
        _items.Add(item);
    }

    public void Remove(Guid id)
    {
        var existing = _items.FirstOrDefault(item => item.Id == id);
        if (existing is null) return;

        Checkpoint();
        _items.Remove(existing);
        if (SelectedId == id) SelectedId = null;
    }

    public AnnotationItem? Find(Guid id) => _items.FirstOrDefault(item => item.Id == id);

    /// <summary>Topmost item whose hit area contains the point.</summary>
    public AnnotationItem? HitTest(Point point) =>
        _items.AsEnumerable().Reverse().FirstOrDefault(item => item.HitBounds().Contains(point));

    /// <summary>Replaces an item in place, without a new checkpoint.</summary>
    public void Replace(AnnotationItem item)
    {
        var index = _items.FindIndex(existing => existing.Id == item.Id);
        if (index >= 0) _items[index] = item;
    }

    public void Clear()
    {
        if (_items.Count == 0) return;
        Checkpoint();
        _items.Clear();
        SelectedId = null;
    }

    public void Undo()
    {
        if (_undo.Count == 0) return;
        _redo.Push(_items.Select(item => item.Clone()).ToList());
        Restore(_undo.Pop());
    }

    public void Redo()
    {
        if (_redo.Count == 0) return;
        _undo.Push(_items.Select(item => item.Clone()).ToList());
        Restore(_redo.Pop());
    }

    private void Restore(List<AnnotationItem> state)
    {
        _items.Clear();
        _items.AddRange(state);
        if (SelectedId is not null && _items.All(item => item.Id != SelectedId)) SelectedId = null;
    }

    public JsonArray ToJson() => new(_items.Select(item => (JsonNode)item.ToJson()).ToArray());
}
